Implement a more robust way to avoid infinite generic specialization loops

swiftix opened the pull request:
The existing simple mechanism for avoiding infinite generic specialization loops is based on checking the structural depth and width of types passed as generic type parameters. If the depth or the width of a type is above a certain threshold, the type is considered too complex for generic specialization and no specialization is produced. While this approach prevents the possibility of producing an infinite number of generic specializations for ever-growing generic type parameters, it catches the issue too late in some cases, leading to excessive CPU and memory usage.

Therefore, the new method tries to solve the problem at its root. An infinite generic specialization loop can be triggered by specializing a given generic call-site if and only if:
-  Doing so would result in a loop inside the specialization graph represented by the `GenericSpecializationInformations`, i.e. it would produce direct or indirect recursion involving a generic call
-  The substitutions used by the current generic call-site are structurally more complex than the substitutions used by the same call-site in the previous iteration inside specialization graph. More complex in this context means that the new generic type parameter structurally contains the generic type parameter from a previous iteration inside the specialization graph and has greater structural depth, e.g. `Array<Int>` is more complex than `Int`.

The generic specializer now records all the required information about specializations it produces and uses it later to detect and prevent any generic specializations which would result in an infinite specialization loop. It detects them as early as possible and thus reduces compile times, memory consumption and potentially also reduces the code-size by not generating useless specializations.

Fixes rdar://33214438

swiftix committed:
Define a GenericSpecializationInformation class which can be used to track the history of generic specializations

GenericSpecializationInformation contains information regarding how a given specialized function was created, e.g. which caller function triggered this specialization, which substitutions were used, etc. Provide some debugging flags to dump the collected specialization information.

The information about generic specializations is referenced by the specialized functions and by call-sites originating from specialized functions.

This information can be created/used by the generic specializer to detect generic call-sites whose specialization would result in non-terminating sequence of subsequent generic specializations.

swiftix committed:
Make it possible to mark SILFunction as referenced by metainformation

This marker can be used e.g. when a function is referenced from the generic specialization information (or any other kind of metadata) to indicate that this SIL function should not be completely eliminated by the dead function elimination pass even if it is not referenced by any function_ref instruction anymore. Such a function will be preserved in a zombie list instead if it needs to be eliminated.

swiftix committed:
Implement a more robust way to avoid infinite generic specialization loops

The existing simple mechanism for avoiding infinite generic specialization loops is based on checking the structural depth and width of types passed as generic type parameters. If the depth or the width of a type is above a certain threshold, the type is considered too complex for generic specialization and no specialization is produced. While this approach prevents the possibility of producing an infinite number of generic specializations for ever-growing generic type parameters, it catches the issue too late in some cases, leading to excessive CPU and memory usage.

Therefore, the new method tries to solve the problem at its root. An infinite generic specialization loop can be triggered by specializing a given generic call-site if and only if:
-  Doing so would result in a loop inside the specialization graph represented by the `GenericSpecializationInformations`, i.e. it would produce direct or indirect recursion involving a generic call
-  The substitutions used by the current generic call-site are structurally more complex than the substitutions used by the same call-site in the previous iteration inside specialization graph. More complex in this context means that the new generic type parameter structurally contains the generic type parameter from a previous iteration inside the specialization graph and has greater structural depth, e.g. `Array<Int>` is more complex than `Int`.

The generic specializer now records all the required information about specializations it produces and uses it later to detect and prevent any generic specializations which would result in an infinite specialization loop. It detects them as early as possible and thus reduces compile times, memory consumption and potentially also reduces the code-size by not generating useless specializations.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke benchmark

swiftix commented:
@DougGregor Could you have a look? You are mostly interested in changes found in `Generics.cpp`, because the actual loop detection and substitutions comparison logic is there.

eeckstein commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -166,6 +173,118 @@ static bool isTypeTooComplex(Type t) {
+            Sub.getReplacement()->dump();
+          });
+
+    if (CurSpecializationInfo->getParent()->getName() == GenericFuncName) {
```
Why comparing the name and not the function pointer itself?

eeckstein commented on lib/SIL/SILInstructions.cpp:
```diff
@@ -2223,3 +2227,49 @@ void KeyPathInst::dropReferencedPattern() {
+  if (F == &BuilderF) {
+    return Apply.getSpecializationInfo();
+  }
+
```
You can add a comment here that the following lines are in case of inlining.

swiftix replied:
Fixed.

eeckstein commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -1001,6 +1029,9 @@ class ApplyInstBase<Impl, Base, false> : public Base {
   /// The type of the callee with our substitutions applied.
   SILType SubstCalleeType;
 
+  /// Information about specialization and inlining of this apply.
```
You should add a comment that this is only != null if the apply was inlined. And in this case it points to the specialization info of the inlined function.

swiftix replied:
Fixed

eeckstein commented on test/SILOptimizer/generic_specialization_loops_detection_with_loops.swift:
```diff
@@ -0,0 +1,105 @@
+// RUN: %target-swift-frontend -O -emit-sil -Xllvm -sil-print-generic-specialization-loops -Xllvm -sil-print-generic-specialization-info %s 2>&1 | %FileCheck --check-prefix=CHECK %s
+
```
Is it guaranteed that this test also tests the inlining case? Maybe you could add an inline(__always), just to be sure.
(And BTW, this is a nice test case!)

swiftix replied:
I'm not sure about what you mean here. The functions are mutually recursive, which means that even with `@inline(__always)` the performance inliner woulds stop recursive inlining.

But coming back to your concern regarding the inlining actually happening. There is a line like this in the test:
`Check that the compiler has produced a specialization information for a call-site.` It checks for a specialization information on a call-site, which may only happen if a specialized function was inlined. Is it what you'd like to check? Or do you mean something else?

eeckstein replied:
Yes, that's what I mean. Sorry, I didn't see this check. I think it's good enough.

swiftix replied:
I extended the comment to make it more clear.

eeckstein reviewed (commented):
Nice work!
lgtm
I just have a few minor comments/questions

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

huonw commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -166,6 +173,118 @@ static bool isTypeTooComplex(Type t) {
+/// This is the case if at least one of the substitution type in Subs2 is
+/// "bigger" than the corresponding substitution type in Subs1.
+/// Type T2 is "smaller" than type T1 if T2 is structurally contained in T1.
+static bool growingSubstitutions(SubstitutionList Subs1,
```
I think recursive constraints allow this to be "fooled":

```swift
protocol P {
    associatedtype X: P
}

struct Start {}
struct Step<Param> {}

struct Outer<Param>: P {
    typealias X = Outer<Step<Param>>
}

func foo<T: P>(_: T.Type) {
    foo(T.X.self)
}
foo(Outer<Start>.self)
```

The `T` in `foo` looks like:

- `Outer<Start>`
- `Outer<Step<Start>>`
- `Outer<Step<Step<Start>>>`
- ...

None of which are considered growing because they're never directly contained within one another.

Maybe first pattern-matching through any common prefix (I think it's `TypeMatcher` that does this?) and only then checking for growing-ness of the mismatched sections is enough. Combined with doing this at every level of the call stack, this seems like it should even catch things with arbitrary recursion depth like (variations of):

```swift
struct Outer1<Param>: P {
    typealias X = Outer2<Param>
}
struct Outer2<Param>: P {
    typealias X = Outer3<Param>
}
// ...
struct OuterN<Param>: P {
    typealias X = Outer1<Step<Param>>
}

foo(Outer1<Start>.self)
```

swiftix replied:
@huonw I added support for you use-cases and used `TypeMatcher` for it. Please have a look.

huonw commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -166,6 +173,118 @@ static bool isTypeTooComplex(Type t) {
+    DEBUG(llvm::dbgs() << "Scan caller's specialization history\n");
+  }
+
+  while (CurSpecializationInfo) {
```
It seems like this means that specializing a call stack of depth n is O(n^2)? Are the checks here fast enough that we don't need to worry?

huonw replied:
NB. this loop is what I was thinking of when I said, above:

> Combined with doing this at every level of the call stack

swiftix replied:
I think the checks are fast enough here. 

Let's say we have a specialization chain of depth N. Checking the components if the current callee occurs on this chain is very cheap. And in situations it is not the case and so we are done. If the callee occurs on the specialization chain, then it is potentially a specialization loop and we need to perform more expensive checks to see if substitutions grow. But IMHO it happens only in rare cases and thus it is very unlikely to see a real compilation performance hit here.

swiftix committed:
Support even more complex cases of generic substitution loops

Allow for cases, where the old substitution type `T1` is partially contained in the new substitution type `T2`. Partially contained means that if you drop the common structural "prefix" of `T1` and `T2` and get `T1'` and `T2'` then `T1'` is strictly contained in `T2'`. E.g.  `Outer<Start>` is partially contained in `Outer<Step<Start>>` if you drop the common prefix `Outer`, then `Start` is contained in `Step<Start>`

swiftix commented:
@swift-ci please smoke test

huonw commented on lib/SILOptimizer/Utils/Generics.cpp:
```diff
@@ -166,6 +174,164 @@ static bool isTypeTooComplex(Type t) {
+      continue;
+    // If the new substitution type is getting smaller, the
+    // substitution type does not grow.
+    if (TypeCmp.isStrictlyContainedIn(Type2, Type1))
```
Could you expand on why this is strict? It seems like one could similarly have the interior of the types shrink, e.g. `Outer<Step<Step<Start>>>` -> `Outer<Step<Start>>` -> `Outer<Start>`?

swiftix replied:
Hmm. You're right. I could use here `isPartiallyContainedIn` as well.

huonw reviewed (commented):
Looks good; I'm still curious about the O(n^2) loop and also the choice of strict vs. partial, but I like it, as long as you think the quadratic behavior won't be too bad.

huonw reviewed (approved):
Whoops, didn't click the "approve" radio box.

DougGregor reviewed (commented):
Generally LGTM with the `TypeMatcher` improvements. Do you want to add some statistics about the # of times this prevents specialization due to recursion, and possibly another statistic for the # of times you hit the fallback depth limiter? Those might be useful if things go wrong.

swiftix commented:
@DougGregor Yes, I can add some stats. It is a good idea.

swiftix committed:
Collect statistics about prevented generic specializations

Count how many generic specializations were prevented due to the possibility of creating an infinite generic specialization loop and due to the complexity of their generic type parameters.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke benchmark

swiftix commented:
@swift-ci please smoke benchmark

swiftix commented:
@swift-ci please smoke benchmark

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

swiftix committed:
Allocate GenericSpecializationInformation using SILModule's memory allocator.

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please smoke test

shahmishal commented:
@swift-ci please benchmark

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test
